### PR TITLE
fix: temporary fix for decrypt message

### PR DESCRIPTION
**Pull request by irvingoujAtDevolution**
To achieve non-allocation in decrypt message and maintain the behavior with Windows SSPI provider, I here by propose this temporary solution. 

## Situation
In our LDAP stack , we have our decrypt code that passes the SecurityBuffer of stream type and SecurityBuffer of data type with 0 buffer into the Decrypt function. The code is shown below
``` c#
                  Span<Helpers.SecBuffer> buffers = stackalloc Helpers.SecBuffer[2];
                    buffers[0].BufferType = (uint)SecBufferType.SECBUFFER_STREAM;
                    buffers[0].cbBuffer = (uint)data.Length;
                    buffers[0].pvBuffer = (IntPtr)dataPtr;

                    buffers[1].BufferType = (uint)SecBufferType.SECBUFFER_DATA;
                    buffers[1].cbBuffer = 0;
                    buffers[1].pvBuffer = IntPtr.Zero;
                    Console.WriteLine("=================BEFORE=======================");
                    Console.WriteLine("buffer[0] type = " + buffers[0].BufferType);
                    Console.WriteLine("buffer[0] cbBuffer = " + buffers[0].cbBuffer);
                    Console.WriteLine("buffer[0] pvBuffer = " + buffers[0].pvBuffer);
                    PrintBuffer(buffers[0].pvBuffer, buffers[0].cbBuffer);

                    Console.WriteLine("buffer[1] type = " + buffers[1].BufferType);
                    Console.WriteLine("buffer[1] cbBuffer = " + buffers[1].cbBuffer);
                    Console.WriteLine("buffer[1] pvBuffer = " + buffers[1].pvBuffer);
                    PrintBuffer(buffers[1].pvBuffer, buffers[1].cbBuffer);
                    SSPI.DecryptMessage(_context, buffers, NextRecvSeqNo());
                    byte[] unwrapped = new byte[buffers[1].cbBuffer];
                    Console.WriteLine("=================AFTER=======================");
                    Console.WriteLine("buffer[0] type = " + buffers[0].BufferType);
                    Console.WriteLine("buffer[0] cbBuffer = " + buffers[0].cbBuffer);
                    Console.WriteLine("buffer[0] pvBuffer = " + buffers[0].pvBuffer);
                    PrintBuffer(buffers[0].pvBuffer, buffers[0].cbBuffer);

                    Console.WriteLine("buffer[1] type = " + buffers[1].BufferType);
                    Console.WriteLine("buffer[1] cbBuffer = " + buffers[1].cbBuffer);
                    Console.WriteLine("buffer[1] pvBuffer = " + buffers[1].pvBuffer);
                     PrintBuffer(buffers[1].pvBuffer, buffers[1].cbBuffer);
                    Console.WriteLine("========================================");
                    Marshal.Copy(buffers[1].pvBuffer, unwrapped, 0, unwrapped.Length);
          

                    return unwrapped;
```

under Windows, we would have result like this
```
=================BEFORE=======================
buffer[0] type = 10
buffer[0] cbBuffer = 160
buffer[0] pvBuffer = 2108939100488
buffer = 05-04-07-FF-00-00-00-1C-00-00-00-00-46-C1-1E-2E-0B-DE-7C-85-6E-14-FF-84-C7-1C-3F-1B-79-A3-94-31-B1-FC-06-6E-30-4D-78-6A-B3-25-A3-76-19-0E-51-92-FC-AB-C2-C8-44-B3-A8-84-55-BF-03-02-E3-70-C8-41-BA-42-6C-63-A5-93-6B-98-A6-F7-BE-2A-C0-33-4A-EC-AC-16-8E-3F-A4-9B-8E-38-82-8F-30-29-65-67-D8-22-BF-45-EC-62-99-42-B0-6D-04-EF-5D-3A-43-6A-AB-E1-06-19-2E-92-4D-AA-4D-DE-C4-4D-28-3A-D6-F2-8F-6D-DB-D2-8C-6C-4C-A1-73-92-0B-59-F4-9C-65-E2-9C-4E-01-A8-AC-85-A3-70-27-F0-D7-F2-39-94-CD-82-E6-26
buffer[1] type = 1
buffer[1] cbBuffer = 0
buffer[1] pvBuffer = 0
=================AFTER=======================
buffer[0] type = 10
buffer[0] cbBuffer = 160
buffer[0] pvBuffer = 2108939100488
buffer = 05-04-07-FF-00-00-00-00-00-00-00-00-46-C1-1E-2E-05-04-07-FF-00-00-00-00-00-00-00-00-46-C1-1E-2E-B1-FC-06-6E-30-4D-78-6A-B3-25-A3-76-E7-54-CD-7D-B0-0A-D2-A1-44-34-9F-30-1B-2D-39-14-30-84-00-00-00-48-02-01-03-64-84-00-00-00-3F-04-00-30-84-00-00-00-37-30-84-00-00-00-31-04-12-6D-73-44-53-2D-50-72-69-6E-63-69-70-61-6C-4E-61-6D-65-31-84-00-00-00-17-04-15-49-54-2D-48-45-4C-50-5C-41-64-6D-69-6E-69-73-74-72-61-74-6F-72-30-84-00-00-00-10-02-01-03-65-84-00-00-00-07-0A-01-00-04-00-04-00
buffer[1] type = 1
buffer[1] cbBuffer = 100
buffer[1] pvBuffer = 2108939100548
buffer = 30-84-00-00-00-48-02-01-03-64-84-00-00-00-3F-04-00-30-84-00-00-00-37-30-84-00-00-00-31-04-12-6D-73-44-53-2D-50-72-69-6E-63-69-70-61-6C-4E-61-6D-65-31-84-00-00-00-17-04-15-49-54-2D-48-45-4C-50-5C-41-64-6D-69-6E-69-73-74-72-61-74-6F-72-30-84-00-00-00-10-02-01-03-65-84-00-00-00-07-0A-01-00-04-00-04-00
========================================
```

Notice that
1. the size and the pointer of the buffer[0] stays the same
2. buffer[1].pvBuffer is buffer[0].pvBuffer + 60, where 60 is the number of bytes of token in Kerberos.

This PR fixes the second problem, where it sets buffer[1].pvBuffer = buffer[0].pvBuffer. So that the code example as shown above gets to work.

However, since I am not familiar with the inner structure of the bytes it passes in and out, this is best I can do so far, I would need help from @TheBestTvarynka and @pacmancoder for further investigations of the problem. With this fix. the result gonna look like this
```
=================BEFORE=======================
buffer[0] type = 10
buffer[0] cbBuffer = 160
buffer[0] pvBuffer = 1627793917536
buffer = 05-04-07-FF-00-00-00-1C-00-00-00-00-2C-BA-2C-B7-2B-A7-DB-EF-28-9D-83-7D-77-43-92-CB-7A-06-07-99-FA-B8-8D-9A-74-C6-24-6B-81-2E-D1-36-11-F2-2F-8C-1A-13-B5-C4-1F-2D-EF-1E-48-3E-A5-21-D0-11-C0-51-8C-B2-0A-FB-16-EE-1F-E9-1B-AD-87-2B-D4-51-F7-C4-C1-A4-5F-C8-D8-21-96-6E-D4-62-FC-5A-81-5A-A8-1B-C0-63-DF-26-B5-A6-31-C4-B1-57-3E-3B-47-7E-7D-A8-59-18-99-BD-A6-F0-6E-DA-BE-05-9A-43-B5-61-C4-50-1A-20-93-82-58-B3-FA-62-B4-27-E0-5F-AC-3A-6D-FE-A3-2F-EF-B3-CA-5D-F5-0A-BB-C8-33-AA-2E-63-D4-5D
buffer[1] type = 1
buffer[1] cbBuffer = 0
buffer[1] pvBuffer = 0
=================AFTER=======================
buffer[0] type = 10
buffer[0] cbBuffer = 100 <<----------- a) This needs to be fixed
buffer[0] pvBuffer = 1627793917536
buffer = 30-84-00-00-00-48-02-01-05-64-84-00-00-00-3F-04-00-30-84-00-00-00-37-30-84-00-00-00-31-04-12-6D-73-44-53-2D-50-72-69-6E-63-69-70-61-6C-4E-61-6D-65-31-84-00-00-00-17-04-15-49-54-2D-48-45-4C-50-5C-41-64-6D-69-6E-69-73-74-72-61-74-6F-72-30-84-00-00-00-10-02-01-05-65-84-00-00-00-07-0A-01-00-04-00-04-00
buffer[1] type = 1
buffer[1] cbBuffer = 100
buffer[1] pvBuffer = 1627793917536
buffer = 30-84-00-00-00-48-02-01-05-64-84-00-00-00-3F-04-00-30-84-00-00-00-37-30-84-00-00-00-31-04-12-6D-73-44-53-2D-50-72-69-6E-63-69-70-61-6C-4E-61-6D-65-31-84-00-00-00-17-04-15-49-54-2D-48-45-4C-50-5C-41-64-6D-69-6E-69-73-74-72-61-74-6F-72-30-84-00-00-00-10-02-01-05-65-84-00-00-00-07-0A-01-00-04-00-04-00
========================================
```
Plus, I suspect by swallowing the bytes caller passed in over an FFI, it creates a memory  leak, in here, each call to decrypt looks like there's 60 bytes missing, if it is heap allocated. This memory-leaking behavior exist before this change.

### CODE REVIEW

File: src/utils.rs
Comment (thenextman) [inline src/utils.rs:255]: nit: "the Windows SSPI" should be just "Windows SSPI"

File: src/utils.rs
Comment (thenextman) [inline src/utils.rs:259]: "This is behaviour of" > "The behaviour of _reusing_ the pointer"

**Comment by thenextman** [comment]
This all looks fine to me, but I add @pacmancoder to take a look for any glaring issue in the code

I understand we're looking for a quick fix to unblock things, so I'll approve this once the formatting / CI etc is happy

File: src/utils.rs
Comment (CBenoit) [inline src/utils.rs:313]: suggestion: Keep the unsafe block as small as possible.

```rust
            let data_slice = unsafe {
                // SAFETY: This is safe because…
                std::slice::from_raw_parts_mut(stream_inner_ptr, stream_inner_size)
            };
            data_buffer.set_data(data_slice)?;
```

**Review by CBenoit — COMMENTED** [review]
cc'ing @TheBestTvarynka

File: src/utils.rs
Comment (irvingoujAtDevolution) [reply src/utils.rs]: updated

File: src/utils.rs
Comment (irvingoujAtDevolution) [reply src/utils.rs]: updated

File: src/utils.rs
Comment (irvingoujAtDevolution) [reply src/utils.rs]: updated

Comment (irvingoujAtDevolution) [reply src/utils.rs]: updated

File: src/utils.rs
Comment (CBenoit) [inline src/utils.rs:311]: nitpick:
```suggestion
            // SAFETY: This is safe because the stream buffer is guaranteed to exist and have the correct size.
            let data_slice = unsafe { std::slice::from_raw_parts_mut(stream_inner_ptr, stream_inner_size) };
```

**Comment by irvingoujAtDevolution** [comment]
Convert To draft, waiting  on changes by @TheBestTvarynka

**Comment by TheBestTvarynka** [comment]
> Plus, I suspect by swallowing the bytes caller passed in over an FFI, it creates a memory leak, in here, each call to decrypt looks like there's 60 bytes missing, if it is heap allocated. This memory-leaking behavior exist before this change.

I assume the following things in my implementation and my previous fixes:

* All `SecurityBuffer`s are part of one heap allocated buffer.
* sspi doesn't allocate any new memory during the decryption but just plays with existing buffers and their types.

Thus, when the caller decides to free this heap buffer, he will do it once and use the pointer to start the memory. So, there should not be any memory leaks. But I anyway will look into it. Maybe I missed something.

> With this fix. the result gonna look like this...

Will it be okay for you if, in the resulting buffers, only the DATA buffer will contain the decrypted data (whereas the STREAM buffer will be empty or contain the irrelevant part of the initial encrypted data)?

**Comment by irvingoujAtDevolution** [comment]
> > Plus, I suspect by swallowing the bytes caller passed in over an FFI, it creates a memory leak, in here, each call to decrypt looks like there's 60 bytes missing, if it is heap allocated. This memory-leaking behavior exist before this change.
> 
> I assume the following things in my implementation and my previous fixes:
> 
> * All `SecurityBuffer`s are part of one heap allocated buffer.
> * sspi doesn't allocate any new memory during the decryption but just plays with existing buffers and their types.
> 
> Thus, when the caller decides to free this heap buffer, he will do it once and use the pointer to start of the memory buffer. So, there should not be any memory leaks. Anyway, I will look into it. Maybe I missed something.
> 
> > With this fix. the result gonna look like this...
> 
> Will it be okay for you if, in the resulting buffers, only the DATA buffer will contain the decrypted data (whereas the STREAM buffer will be empty or contain the irrelevant part of the initial encrypted data)?

I see, then memory leak should not be a problem here. the caller always holds the pointer anyway.

With regards to the second part. It would be ok for me, but ultimately we want to maintain the consistency of behaviors between sspi-rs and Windows SSPI.
it's up to @awakecoding to make the decision.

**Comment by awakecoding** [comment]
I'm off work today and didn't re-read the whole exchange, but we're trying to mimic the Windows SSPI behavior as much as possible, with compatibility for the various ways applications can call the Windows SSPI. Sometimes this is difficult but we should do the work

**Comment by pacmancoder** [comment]
I think I would vote to fix the first problem (`the size and the pointer of the buffer[0] stays the same`) right away in this PR, but it depends on the urgency of the fix (?)

Also, when the issue will be fixed completely, It would be great to add the before/after results with Windows SSPI provider as a test case for this function (talking about this one) to avoid regressions in future:

```
=================BEFORE=======================
buffer[0] type = 10
buffer[0] cbBuffer = 160
buffer[0] pvBuffer = 2108939100488
buffer = 05-04-07-FF-00-00-00-1C-00-00-00-00-46-C1-1E-2E-0B-DE-7C-85-6E-14-FF-84-C7-1C-3F-1B-79-A3-94-31-B1-FC-06-6E-30-4D-78-6A-B3-25-A3-76-19-0E-51-92-FC-AB-C2-C8-44-B3-A8-84-55-BF-03-02-E3-70-C8-41-BA-42-6C-63-A5-93-6B-98-A6-F7-BE-2A-C0-33-4A-EC-AC-16-8E-3F-A4-9B-8E-38-82-8F-30-29-65-67-D8-22-BF-45-EC-62-99-42-B0-6D-04-EF-5D-3A-43-6A-AB-E1-06-19-2E-92-4D-AA-4D-DE-C4-4D-28-3A-D6-F2-8F-6D-DB-D2-8C-6C-4C-A1-73-92-0B-59-F4-9C-65-E2-9C-4E-01-A8-AC-85-A3-70-27-F0-D7-F2-39-94-CD-82-E6-26
buffer[1] type = 1
buffer[1] cbBuffer = 0
buffer[1] pvBuffer = 0
=================AFTER=======================
buffer[0] type = 10
buffer[0] cbBuffer = 160
buffer[0] pvBuffer = 2108939100488
buffer = 05-04-07-FF-00-00-00-00-00-00-00-00-46-C1-1E-2E-05-04-07-FF-00-00-00-00-00-00-00-00-46-C1-1E-2E-B1-FC-06-6E-30-4D-78-6A-B3-25-A3-76-E7-54-CD-7D-B0-0A-D2-A1-44-34-9F-30-1B-2D-39-14-30-84-00-00-00-48-02-01-03-64-84-00-00-00-3F-04-00-30-84-00-00-00-37-30-84-00-00-00-31-04-12-6D-73-44-53-2D-50-72-69-6E-63-69-70-61-6C-4E-61-6D-65-31-84-00-00-00-17-04-15-49-54-2D-48-45-4C-50-5C-41-64-6D-69-6E-69-73-74-72-61-74-6F-72-30-84-00-00-00-10-02-01-03-65-84-00-00-00-07-0A-01-00-04-00-04-00
buffer[1] type = 1
buffer[1] cbBuffer = 100
buffer[1] pvBuffer = 2108939100548
buffer = 30-84-00-00-00-48-02-01-03-64-84-00-00-00-3F-04-00-30-84-00-00-00-37-30-84-00-00-00-31-04-12-6D-73-44-53-2D-50-72-69-6E-63-69-70-61-6C-4E-61-6D-65-31-84-00-00-00-17-04-15-49-54-2D-48-45-4C-50-5C-41-64-6D-69-6E-69-73-74-72-61-74-6F-72-30-84-00-00-00-10-02-01-03-65-84-00-00-00-07-0A-01-00-04-00-04-00
========================================
```

**Comment by irvingoujAtDevolution** [comment]
@TheBestTvarynka @thenextman Do you guys mind if we merge this PR as temporary fix for now? Then we push the real fix later. @pauldumais needs it to fix our LDAP stack in RDM.

**Comment by thenextman** [comment]
@pacmancoder @TheBestTvarynka @irvingoujAtDevolution The issue is important and blocking us on RDM side, unless there is something I missed I would vote to merge this now and address the proper fix at a high priority.

**Comment by TheBestTvarynka** [comment]
I'm ok if you merge it today. I'll provide my fix tomorrow (I need to pass the internal review firss and check that my code didn't break anything else)

**Comment by thenextman** [comment]
> I'm ok if you merge it today. I'll provide my fix tomorrow (I need to pass the internal review firss and check that my code didn't break anything else)

Thanks!

**Review by thenextman — APPROVED** [review]

**Pull request merged**
